crates/ruff_python_formatter/README.md

```diff
@@ -323,3 +323,188 @@ await [1, 2, 3]
 This is more consistent to the formatting of other awaited expressions: Ruff and Black both
 remove parentheses around, e.g., `await (1)`, only retaining them when syntactically required,
 as in, e.g., `await (x := 1)`.
+
+### Line breaks between implicitly concatenated string attributes are not preserved ([#7052](https://github.com/astral-sh/ruff/issues/7052))
+
+Black preserves line breaks between implicitly concatenated string attributes.
```

Comment:
I don't think black is preserving the line breaks. What black does is it tries to fit the entire logical line:

```python
print("aaaaaaa {:.2f}% " "aaaaaaaaaaaaaaaa".format((((df_aaaaaaaaaaa["key"].isna()) | (df_aaaaaaaaaaa["key"] == 0)).sum() / len(df_aaaaaaaaaaa)) * 100))
```

Which is way too long. First, it expands the outermost call

```python
print(
    "aaaaaaa {:.2f}% " "aaaaaaaaaaaaaaaa".format((((df_aaaaaaaaaaa["key"].isna()) | (df_aaaaaaaaaaa["key"] == 0)).sum() / len(df_aaaaaaaaaaa)) * 100)
)
```

But the argument line is still too long. So black searches for the operator with the highest split precedence which in this case is the implicit string concatenation and it splits between the parts

```python
print(
    "aaaaaaa {:.2f}% "
    "aaaaaaaaaaaaaaaa".format((((df_aaaaaaaaaaa["key"].isna()) | (df_aaaaaaaaaaa["key"] == 0)).sum() / len(df_aaaaaaaaaaa)) * 100)
)
```

This makes the first argument-line fit, but not the second. So black continues splitting that one by iteratively searching the highest precedence operator in the line and splitting around all operators with the given priority. 

I think the "easiest" explanation is that Black gives implicit concatenated a high split priority which makes sense in binary like expressions but IMO leads to undesired results in attribute chains.

Reply:
Thanks, this is tricky but I did my best. I appreciate the clarification here.